### PR TITLE
List kinesis

### PR DESCRIPTION
**Pull request by efimk-lu**
Testing are not ready yet, but you can start to give initial feedback.
I'm puzzled on the columns which the table should have.

### CODE REVIEW

File: src/commands/list-kinesis-streams.js [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+			ExclusiveStartStreamName: streamDetails.slice(-1)[0]
+		}).promise();
+		streamNames = streamDetails.StreamNames;
+	}
+	const usageStats = await getUsageMetrics(streamNames, region);
+	return describeStreams(streamNames, region, usageStats);
+};
+
+const getStreamsDescriptionFromAllRegions = async () => {
+	const promises = Lambda.regions.map(region =>
```
Comment (theburningmonk) [inline src/commands/list-kinesis-streams.js]: Kinesis probably has a different list of available regions

File: src/commands/list-kinesis-streams.js [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+		description: "AWS CLI profile name",
+		required: false
+	})
+};
+
+const getStreamsDescriptionFromRegion = async region => {
+	const Kinesis = new AWS.Kinesis({ region });
+	let streamDetails = await Kinesis.listStreams({ Limit: 100 }).promise();
+	let streamNames = streamDetails.StreamNames;
+	while (streamDetails.HasMoreStreams) {
```
Comment (theburningmonk) [inline src/commands/list-kinesis-streams.js]: I don't think this works, this would only return the stream details and stream names from the last batch

Comment (theburningmonk) [reply src/commands/list-kinesis-streams.js]: I have some example where I use recursion for this type of thing, probably in the list lambda cmd

File: src/commands/list-kinesis-streams.js [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+		["IncomingBytes", "Sum"],
+		["IncomingRecords", "Sum"],
+		["GetRecords.Bytes", "Sum"],
+		["GetRecords.Records", "Sum"]
+	];
+
+	const oneMinuteAgo = new Date();
+	oneMinuteAgo.setMinutes(oneMinuteAgo.getMinutes() - 1);
+
+	const queries = _.flatMap(streamNames, streamName =>
```
Comment (theburningmonk) [inline src/commands/list-kinesis-streams.js]: you'd need to chunk this, CW only allow you to run X number of queries per request

Comment (theburningmonk) [reply src/commands/list-kinesis-streams.js]: nvm, just saw you're doing it later

File: src/commands/list-kinesis-streams.js [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+		const kvp = shardMetrics.map(({ metricName, dataPoint }) => [
+			metricName,
+			dataPoint
+		]);
+		return _.fromPairs(kvp);
+	});
+};
+
+const show = streamsDetails => {
+	const table = new Table({
```
Comment (theburningmonk) [inline src/commands/list-kinesis-streams.js]: these column names are too long, can we shorten them? e.g. `# of Shard` => `Shards`, `Avg Kinesis Read Capacity (MB/Records) %` => `Read utilization`, `Avg Kinesis Write Capacity (MB/Records) %`> `Write utilization`?

if you feel these need more explanation, then let's add some description at the bottom of the table

File: src/commands/list-kinesis-streams.js [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+		head: [
+			"Stream name",
+			"ARN",
```
Comment (theburningmonk) [inline src/commands/list-kinesis-streams.js]: I don't think we need `ARN` but it'll be useful to have `region` in here since you can list all streams in all regions

File: src/commands/list-kinesis-streams.js [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+		StreamName: streamName
+	}).promise();
+
+	return {
+		streamName: streamName,
+		arn: resp.StreamDescription.StreamARN,
+		status: resp.StreamDescription.StreamStatus,
+		shards: resp.StreamDescription.Shards,
+		readMbPercentage: formatFloat(
+			(streamStat["GetRecords.BytesSum"] /
```
Comment (theburningmonk) [inline src/commands/list-kinesis-streams.js]: this calculation is kinda hard to follow, maybe extract it into a method and use a couple of intermediate variables so one can follow the steps of the calculation easier?

File: src/commands/list-kinesis-streams.js [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+		getStreamsDescriptionFromRegion(region)
+	);
+	const results = await Promise.all(promises);
+	return _.flatMap(results);
+};
+
+const describeStreams = async (streams, region, usageStats) => {
+	const streamsDescription = [];
+	for (const stream of streams) {
+		streamsDescription.push(await describeStream(stream, region, usageStats[stream]));
```
Comment (theburningmonk) [inline src/commands/list-kinesis-streams.js]: can this be done in parallel rather than one at a time?

File: src/commands/list-kinesis-streams.js [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+};
+
+class ListKinesisStreamsCommand extends Command {
+	async run() {
+		const { flags } = this.parse(ListKinesisStreamsCommand);
+		const { region, profile } = flags;
+
+		AWS.config.region = region;
+		if (profile) {
+			AWS.config.credentials = new AWS.SharedIniFileCredentials({ profile });
```
Comment (theburningmonk) [inline src/commands/list-kinesis-streams.js]: there's a `lib/aws` module, which saves you some hassle for this, see https://github.com/lumigo-io/lumigo-CLI/blob/master/src/commands/list-lambda.js for how I'm using it in other places

File: src/commands/list-kinesis-streams.js [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+const getUsageMetrics = async (streamNames, region) => {
+	const CloudWatch = new AWS.CloudWatch({ region });
+	const metricNames = [
+		["IncomingBytes", "Sum"],
+		["IncomingRecords", "Sum"],
+		["GetRecords.Bytes", "Sum"],
+		["GetRecords.Records", "Sum"]
+	];
+
+	const oneMinuteAgo = new Date();
```
Comment (theburningmonk) [inline src/commands/list-kinesis-streams.js]: should we use the last 5/10 mins instead of just the last minute?

File: src/commands/list-kinesis-streams.js (hunk 1 above)
Comment (efimk-lu) [reply src/commands/list-kinesis-streams.js]: Done

File: src/commands/list-kinesis-streams.js (hunk 2 above)
Comment (efimk-lu) [reply src/commands/list-kinesis-streams.js]: I don't fully understand why we need recursion, it works well with a loop

File: src/commands/list-kinesis-streams.js (hunk 4 above)
Comment (efimk-lu) [reply src/commands/list-kinesis-streams.js]: Done

File: src/commands/list-kinesis-streams.js (hunk 5 above)
Comment (efimk-lu) [reply src/commands/list-kinesis-streams.js]: Done

File: src/commands/list-kinesis-streams.js (hunk 6 above)
Comment (efimk-lu) [reply src/commands/list-kinesis-streams.js]: Done

File: src/commands/list-kinesis-streams.js (hunk 8 above)
Comment (efimk-lu) [reply src/commands/list-kinesis-streams.js]: Done

File: src/commands/list-kinesis-streams.js (hunk 9 above)
Comment (efimk-lu) [reply src/commands/list-kinesis-streams.js]: 5 min it is

File: src/commands/list-kinesis-streams.js [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+	}),
+	profile: flags.string({
+		char: "p",
+		description: "AWS CLI profile name",
+		required: false
+	})
+};
+
+const getStreamsDescriptionFromRegion = async region => {
+	let streamNames = await getStreamsInRegion(region);
```
Comment (theburningmonk) [inline src/commands/list-kinesis-streams.js]: `const` instead of `let`?

File: src/commands/list-kinesis-streams.js [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+	);
+	const results = await Promise.all(promises);
+	return _.flatMap(results);
+};
+
+const describeStreams = async (streams, region, usageStats) => {
+	const streamsDescription = [];
+	for (const stream of streams) {
+		streamsDescription.push(describeStream(stream, region, usageStats[stream]));
+	}
```
Comment (theburningmonk) [inline src/commands/list-kinesis-streams.js]: not an issue but it's probably nicer as a `map`

File: src/commands/list-kinesis-streams.js [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+
+	const incomingBytesForAllShardsPerFiveMinute =
+		numberOfShards * BYTES_INCOMING_PER_SHARD * FIVE_MIN_IN_SECONDS;
+
+	const incomingRecordsForAllShardsPerFiveMinute =
+		numberOfShards * MAX_RECORDS_PER_SHARD_WRITE * FIVE_MIN_IN_SECONDS;
+
+	return {
+		streamName: streamName,
+		region: getRegionFromARN(resp.StreamDescription.StreamARN),
```
Comment (theburningmonk) [inline src/commands/list-kinesis-streams.js]: you already had the `region` as function argument

File: src/commands/list-kinesis-streams.js [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+			"Status"
+		]
+	});
+
+	streamsDetails.forEach(x => {
+		table.push([
+			x.streamName,
+			x.region,
+			x.shards.length,
+			`${x.outgoingMbPercentage}% (MB) \n${x.outgoingRecordPercentage}% (Records)`,
```
Comment (theburningmonk) [inline src/commands/list-kinesis-streams.js]: redundant trailing whitespace before `\n`?

File: test/commands/list-kinesis-streams.test.js [hunk 14]
```diff
@@ -0,0 +1,108 @@
+const { expect, test } = require("@oclif/test");
```
Comment (theburningmonk) [inline test/commands/list-kinesis-streams.test.js]: the indentation in this file seems completely different to the rest of the project for some reason, is it using tabs instead of spaces?

Comment (efimk-lu) [reply test/commands/list-kinesis-streams.test.js]: @theburningmonk I think so, that's what prettier gave me, do you want me to change it?

File: src/commands/list-kinesis-streams.js (hunk 11 above)
Comment (efimk-lu) [reply src/commands/list-kinesis-streams.js]: I'll ignore gracefully ;-)

**Comment by efimk-lu** [comment]
BTW @theburningmonk I've noticed that some tests fail due to the check of the latest version

File: src/commands/list-kinesis-streams.js (hunk 7 above)
Comment (efimk-lu) [reply src/commands/list-kinesis-streams.js]: It's done in parallel using `Promise.all(streamsDescription)`
